technologies/FlexBE/pyrobosim_flexbe_btcpp/src/flexbe_btcpp_executor.cpp

```diff
@@ -0,0 +1,221 @@
+  {
+    RCLCPP_INFO(kLogger, "onTreeExecutionCompleted with status=%d (canceled=%d) after %d tics",
+                int(status), was_cancelled, tic_count_);
+    logger_cout_.reset();
+    std::string result =
+        "pyrobosim_flexbe_btcpp tree completed with status=" + std::to_string(int(status)) +
+        " after " + std::to_string(tic_count_) + " tics";
+
+    return result;
+  }
+
+  /**
+   * @brief Callback invoked when a goal is received and before the tree is created.
+   * If it returns false, the goal will be rejected.
+  */
+  bool onGoalReceived(const std::string& tree_name, const std::string& payload)
+  {
+    RCLCPP_INFO(kLogger, "onGoalReceived with tree name '%s' with payload '%s'", tree_name.c_str(),
+                payload.c_str());
+    return true;
+  }
+
+  /**
+   * @brief registerNodesIntoFactory is a callback invoked after the
+   * plugins were registered into the BT::BehaviorTreeFactory.
+   * It can be used to register additional custom nodes manually.
+   *
+   * @param factory The factory to use to register nodes
+  */
+  void registerNodesIntoFactory(BT::BehaviorTreeFactory& factory)
+  {
+    RCLCPP_INFO(kLogger, "registerNodesIntoFactory - load custom nodes ...");
+    // all the actions are done using the same Action Server.
+    // Therefore  single RosNodeParams will do.
+    BT::RosNodeParams params;
+    params.nh = this->node();
+    params.default_port_value = "execute_action";
+
+    factory.registerNodeType<BT::CloseAction>("Close", params);
+    factory.registerNodeType<BT::DetectObject>("DetectObject", params);
+    factory.registerNodeType<BT::NavigateAction>("Navigate", params);
+    factory.registerNodeType<BT::OpenAction>("Open", params);
+    factory.registerNodeType<BT::PickObject>("PickObject", params);
+    factory.registerNodeType<BT::PlaceObject>("PlaceObject", params);
+    RCLCPP_INFO(kLogger, "registerNodesIntoFactory - registered custom nodes.");
+  }
+
+  /**
+   * @brief onLoopAfterTick invoked at each loop, after tree.tickOnce().
+   * If it returns a valid NodeStatus, the tree will stop and return that status.
+   * Return std::nullopt to continue the execution.
+   *
+   * @param status The status of the tree after the last tick
+  */
+  std::optional<BT::NodeStatus> onLoopAfterTick(BT::NodeStatus /*status*/)
+  {
+    //RCLCPP_INFO(kLogger, "pyrobosim_flexbe_btcpp - onLoopAfterTick.");
+    ++tic_count_;
+    return std::nullopt;
+  }
+
+  /**
+   * @brief onLoopFeedback is a callback invoked at each loop, after tree.tickOnce().
+   * If it returns a valid string, it will be sent as feedback to the Action Client.
+   *
+   * If you don't want to return any feedback, return std::nullopt.
+  */
+  std::optional<std::string> onLoopFeedback()
+  {
+    return std::nullopt;
+  }
+
+private:
+  std::shared_ptr<BT::StdCoutLogger> logger_cout_;
+  rclcpp::Subscription<pyrobosim_msgs::msg::RobotState>::SharedPtr sub_;
```

Comment:
nit: personal preference, but I like to call these variables by what they subscribe to, e.g., `state_sub_` or `robot_state_sub_`.